icechunk/src/storage/mod.rs

```diff
@@ -323,28 +324,38 @@ where
 
 /// Split an object request into multiple byte range requests
 ///
-/// Returns tuples of Range for each request. It tries to generate the maximum number of
-/// requests possible, not generating more than `max_parts` requests, and each request not being
-/// smaller than `min_part_size`. Note that the size of the last request is >= the preceding one.
+/// Returns tuples of Range for each request.
+///
+/// It generates requests that are as similar as possible in size, this means no more than 1 byte
+/// difference between the requests.
+///
+/// It tries to generate ceil(size/ideal_req_size) requests, but never exceeds max_requests.
+///
+/// ideal_req_size and max_requests must be > 0
 fn split_in_multiple_requests(
     range: &Range<u64>,
-    min_part_size: u64,
-    max_parts: u16,
+    ideal_req_size: u64,
+    max_requests: u16,
 ) -> impl Iterator<Item = Range<u64>> {
-    let size = range.end - range.start;
-    let min_part_size = max(min_part_size, 1);
-    let num_parts = size / min_part_size;
-    let num_parts = max(1, min(num_parts, max_parts as u64));
-    let equal_parts = num_parts - 1;
-    let equal_parts_size = size / num_parts;
-    let last_part_size = size - equal_parts * equal_parts_size;
-    let equal_requests = iter::successors(
-        Some(range.start..range.start + equal_parts_size),
-        move |range| Some(range.end..range.end + equal_parts_size),
-    );
-    let last_part_offset = range.start + equal_parts * equal_parts_size;
-    let last_request = iter::once(last_part_offset..last_part_offset + last_part_size);
-    equal_requests.take(equal_parts as usize).chain(last_request)
+    let size = max(0, range.end - range.start);
+    // we do a ceiling division, rounding always up
+    let num_parts = (size - 1) / ideal_req_size + 1;
```

Comment:
v.minor: could use the built-in method for clarity

```suggestion
    let num_parts = (size - 1).div_ceil(ideal_req_size);
```
https://doc.rust-lang.org/std/primitive.u64.html#method.div_ceil

Reply:
What? is that a thing? i swear i searched for it ... with the wrong name i guess

Reply:
done